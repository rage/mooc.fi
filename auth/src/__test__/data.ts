import { NextFunction, Request, Response } from "express"

export const testProfile = {
  edu_person_affiliation: "",
  edu_person_principal_name: "test",
  schac_home_organization: "helsinki.fi",
  organizational_unit: "organizational_unit",
  schac_personal_unique_code:
    "urn:schac:personalUniqueCode:int:studentID:helsinki.fi:121345678",
  display_name: "test",
  given_name: "test",
  mail: "test@mail.com",
}

export const req = {
  body: {},
  query: {
    RelayState: JSON.stringify({
      language: "fi",
      provider: "hy",
      action: "sign-in",
    }),
  },
  params: {},
  login: jest.fn().mockImplementation((_: any, fn: Function) => fn(null)),
  logout: jest.fn(),
} as unknown as Request
export const res = {
  redirect: jest.fn().mockReturnThis(),
  locals: {},
  setMOOCCookies: jest.fn().mockReturnThis(),
} as unknown as Response
export const next = jest.fn() as NextFunction

export const okTokenResponse = {
  data: {
    status: 200,
    success: true,
    tmc_token: "tmc_token",
    access_token: "access_token",
    admin: false,
  },
}
