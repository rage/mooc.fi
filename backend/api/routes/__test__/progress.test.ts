import {
  createRequestHelpers,
  fakeTMCCurrent,
  getTestContext,
  RequestGet,
} from "../../../tests"
import { normalUserDetails, seed } from "../../../tests/data"

const ctx = getTestContext()

describe("API", () => {
  describe("/api/user-course-progress", () => {
    const tmc = fakeTMCCurrent({
      "Bearer normal": [200, normalUserDetails],
      "Bearer third": [200, { ...normalUserDetails, id: 3 }],
    })

    beforeAll(() => tmc.setup())
    afterAll(() => tmc.teardown())

    let getUserCourseProgress: (slug: string) => RequestGet
    beforeEach(async () => {
      await seed(ctx.prisma)
      const { get } = createRequestHelpers(ctx.port)
      getUserCourseProgress = (slug: string) =>
        get(`/api/user-course-progress/${slug}`, {})
    })

    it("errors on no auth", async () => {
      return getUserCourseProgress("course1")({})
        .then(() => fail())
        .catch(({ response }) => {
          expect(response.status).toBe(401)
        })
    })

    it("returns data", async () => {
      const ucpResponse = await getUserCourseProgress("course1")({
        headers: {
          Authorization: "Bearer normal",
        },
      })

      expect(ucpResponse.status).toBe(200)
      expect(ucpResponse.data).not.toBeNull()

      expect(ucpResponse.data.data.progress).toEqual([
        { group: "week1", max_points: 3, n_points: 0 },
      ])
    })

    it("returns only the oldest instance", async () => {
      const ucpResponse = await getUserCourseProgress("course1")({
        headers: {
          Authorization: "Bearer third",
        },
      })

      expect(ucpResponse.status).toBe(200)
      expect(ucpResponse.data).not.toBeNull()

      expect(ucpResponse.data.data.progress).toEqual([
        { group: "week1", max_points: 3, n_points: 3 },
      ])
      expect(ucpResponse.data.data.created_at).toEqual(
        "1900-01-01T08:00:00.000Z",
      )
    })
  })

  describe("/api/progressv2", () => {
    const tmc = fakeTMCCurrent({
      "Bearer normal": [200, normalUserDetails],
      "Bearer third": [200, { ...normalUserDetails, id: 3 }],
    })

    beforeAll(() => tmc.setup())
    afterAll(() => tmc.teardown())

    let getProgressv2: (idOrSlug: string, deleted?: boolean) => RequestGet

    beforeEach(async () => {
      await seed(ctx.prisma)
      const { get } = createRequestHelpers(ctx.port)
      getProgressv2 = (idOrSlug: string, deleted: boolean = false) =>
        get(`/api/progressv2/${idOrSlug}${deleted ? "?deleted=true" : ""}`, {})
    })

    it("errors on no auth", async () => {
      return getProgressv2("course1")({})
        .then(() => fail())
        .catch(({ response }) => {
          expect(response.status).toBe(401)
        })
    })

    it("errors on no course found", async () => {
      return getProgressv2("foo")({
        headers: { Authorization: "Bearer normal" },
      })
        .then(() => fail())
        .catch(({ response }) => {
          expect(response.status).toBe(404)
          expect(response.data.message).toContain("course not found")
        })
    })

    test.each([
      {
        paramType: "id",
        param: "00000000000000000000000000000001",
        deleted: false,
      },
      { paramType: "slug", param: "course2", deleted: false },
      {
        paramType: "id",
        param: "00000000000000000000000000000001",
        deleted: true,
      },
      { paramType: "slug", param: "course2", deleted: true },
    ])(
      "returns correct data on course $paramType, include deleted $deleted",
      async ({ param, deleted }) => {
        const res = await getProgressv2(
          param,
          deleted,
        )({
          headers: { Authorization: "Bearer normal" },
        })

        expect(res.data).toMatchSnapshot()
      },
    )

    // TODO/FIXME: maybe should also test if it returns handler completion,
    // but similar functions are already tested
  })
})
