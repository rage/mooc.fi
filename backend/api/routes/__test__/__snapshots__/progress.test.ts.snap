// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`API /api/progressv2 returns correct data on course id, include deleted false 1`] = `
{
  "data": {
    "completion": {
      "certificate_id": null,
      "completion_date": "2000-01-01T08:00:00.000Z",
      "completion_language": null,
      "completion_registration_attempt_date": "2022-01-01T08:00:00.000Z",
      "course_id": "00000000-0000-0000-0000-000000000001",
      "created_at": "2000-01-01T08:00:00.000Z",
      "eligible_for_ects": true,
      "email": "e@mail.com",
      "grade": null,
      "id": "30000000-0000-0000-0000-000000000103",
      "student_number": null,
      "tier": null,
      "updated_at": "2000-01-01T08:00:00.000Z",
      "user_id": "20000000-0000-0000-0000-000000000102",
      "user_upstream_id": 1,
    },
    "course_id": "00000000-0000-0000-0000-000000000001",
    "exercise_completions": {
      "50000000-0000-0000-0000-000000000001": {
        "attempted": false,
        "completed": true,
        "exercise_id": "50000000-0000-0000-0000-000000000001",
        "max_points": 2,
        "n_points": "6.000000000000000000000000000000",
        "part": 1,
        "quizzes_id": "customid1",
        "section": 1,
        "user_id": "20000000-0000-0000-0000-000000000102",
      },
      "50000000-0000-0000-0000-000000000002": {
        "attempted": false,
        "completed": false,
        "exercise_id": "50000000-0000-0000-0000-000000000002",
        "max_points": 3,
        "n_points": "0.000000000000000000000000000000",
        "part": 1,
        "quizzes_id": "customid2",
        "section": 2,
        "user_id": "20000000-0000-0000-0000-000000000102",
      },
    },
    "user_id": "20000000-0000-0000-0000-000000000102",
  },
}
`;

exports[`API /api/progressv2 returns correct data on course id, include deleted true 1`] = `
{
  "data": {
    "completion": {
      "certificate_id": null,
      "completion_date": "2000-01-01T08:00:00.000Z",
      "completion_language": null,
      "completion_registration_attempt_date": "2022-01-01T08:00:00.000Z",
      "course_id": "00000000-0000-0000-0000-000000000001",
      "created_at": "2000-01-01T08:00:00.000Z",
      "eligible_for_ects": true,
      "email": "e@mail.com",
      "grade": null,
      "id": "30000000-0000-0000-0000-000000000103",
      "student_number": null,
      "tier": null,
      "updated_at": "2000-01-01T08:00:00.000Z",
      "user_id": "20000000-0000-0000-0000-000000000102",
      "user_upstream_id": 1,
    },
    "course_id": "00000000-0000-0000-0000-000000000001",
    "exercise_completions": {
      "50000000-0000-0000-0000-000000000001": {
        "attempted": false,
        "completed": true,
        "exercise_id": "50000000-0000-0000-0000-000000000001",
        "max_points": 2,
        "n_points": "6.000000000000000000000000000000",
        "part": 1,
        "quizzes_id": "customid1",
        "section": 1,
        "user_id": "20000000-0000-0000-0000-000000000102",
      },
      "50000000-0000-0000-0000-000000000002": {
        "attempted": false,
        "completed": false,
        "exercise_id": "50000000-0000-0000-0000-000000000002",
        "max_points": 3,
        "n_points": "0.000000000000000000000000000000",
        "part": 1,
        "quizzes_id": "customid2",
        "section": 2,
        "user_id": "20000000-0000-0000-0000-000000000102",
      },
      "50000000-0000-0000-0000-000000000005": {
        "attempted": true,
        "completed": false,
        "exercise_id": "50000000-0000-0000-0000-000000000005",
        "max_points": null,
        "n_points": "1.000000000000000000000000000000",
        "part": null,
        "quizzes_id": "customid5",
        "section": null,
        "user_id": "20000000-0000-0000-0000-000000000102",
      },
    },
    "user_id": "20000000-0000-0000-0000-000000000102",
  },
}
`;

exports[`API /api/progressv2 returns correct data on course slug, include deleted false 1`] = `
{
  "data": {
    "completion": {
      "certificate_id": null,
      "completion_date": "2000-01-01T08:00:00.000Z",
      "completion_language": null,
      "completion_registration_attempt_date": "2022-01-01T08:00:00.000Z",
      "course_id": "00000000-0000-0000-0000-000000000001",
      "created_at": "2000-01-01T08:00:00.000Z",
      "eligible_for_ects": true,
      "email": "e@mail.com",
      "grade": null,
      "id": "30000000-0000-0000-0000-000000000103",
      "student_number": null,
      "tier": null,
      "updated_at": "2000-01-01T08:00:00.000Z",
      "user_id": "20000000-0000-0000-0000-000000000102",
      "user_upstream_id": 1,
    },
    "course_id": "00000000-0000-0000-0000-000000000001",
    "exercise_completions": {
      "50000000-0000-0000-0000-000000000001": {
        "attempted": false,
        "completed": true,
        "exercise_id": "50000000-0000-0000-0000-000000000001",
        "max_points": 2,
        "n_points": "6.000000000000000000000000000000",
        "part": 1,
        "quizzes_id": "customid1",
        "section": 1,
        "user_id": "20000000-0000-0000-0000-000000000102",
      },
      "50000000-0000-0000-0000-000000000002": {
        "attempted": false,
        "completed": false,
        "exercise_id": "50000000-0000-0000-0000-000000000002",
        "max_points": 3,
        "n_points": "0.000000000000000000000000000000",
        "part": 1,
        "quizzes_id": "customid2",
        "section": 2,
        "user_id": "20000000-0000-0000-0000-000000000102",
      },
    },
    "user_id": "20000000-0000-0000-0000-000000000102",
  },
}
`;

exports[`API /api/progressv2 returns correct data on course slug, include deleted true 1`] = `
{
  "data": {
    "completion": {
      "certificate_id": null,
      "completion_date": "2000-01-01T08:00:00.000Z",
      "completion_language": null,
      "completion_registration_attempt_date": "2022-01-01T08:00:00.000Z",
      "course_id": "00000000-0000-0000-0000-000000000001",
      "created_at": "2000-01-01T08:00:00.000Z",
      "eligible_for_ects": true,
      "email": "e@mail.com",
      "grade": null,
      "id": "30000000-0000-0000-0000-000000000103",
      "student_number": null,
      "tier": null,
      "updated_at": "2000-01-01T08:00:00.000Z",
      "user_id": "20000000-0000-0000-0000-000000000102",
      "user_upstream_id": 1,
    },
    "course_id": "00000000-0000-0000-0000-000000000001",
    "exercise_completions": {
      "50000000-0000-0000-0000-000000000001": {
        "attempted": false,
        "completed": true,
        "exercise_id": "50000000-0000-0000-0000-000000000001",
        "max_points": 2,
        "n_points": "6.000000000000000000000000000000",
        "part": 1,
        "quizzes_id": "customid1",
        "section": 1,
        "user_id": "20000000-0000-0000-0000-000000000102",
      },
      "50000000-0000-0000-0000-000000000002": {
        "attempted": false,
        "completed": false,
        "exercise_id": "50000000-0000-0000-0000-000000000002",
        "max_points": 3,
        "n_points": "0.000000000000000000000000000000",
        "part": 1,
        "quizzes_id": "customid2",
        "section": 2,
        "user_id": "20000000-0000-0000-0000-000000000102",
      },
      "50000000-0000-0000-0000-000000000005": {
        "attempted": true,
        "completed": false,
        "exercise_id": "50000000-0000-0000-0000-000000000005",
        "max_points": null,
        "n_points": "1.000000000000000000000000000000",
        "part": null,
        "quizzes_id": "customid5",
        "section": null,
        "user_id": "20000000-0000-0000-0000-000000000102",
      },
    },
    "user_id": "20000000-0000-0000-0000-000000000102",
  },
}
`;
